datadog.go

```diff
@@ -0,0 +1,277 @@
+package datadog
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// The maximum content size per request is 5MB
+	maxContentSize = 5 * 1024 * 1024
+
+	// The maximum size for a single log is 256kB
+	maxLogSize = 256 * 1024
+
+	// The maximum amount of logs that can be sent in a single request is 1000
+	maxLogCount = 1000
+
+	basePath          = "/v1/input"
+	datadogHostUS     = "https://http-intake.logs.datadoghq.com"
+	datadogHostEU     = "https://http-intake.logs.datadoghq.eu"
+	apiKeyHeader      = "DD-API-KEY"
+	defaultMaxRetries = 5
+)
+
+var (
+	defaultLevel = logrus.InfoLevel
+)
+
+// These options if provided will be added as default tags to each log sent
+type GlobalTags struct {
+	Service     string
+	Environment string
+	Maintainer  string
+	Application string
+	Hostname    string
+}
+
+type DatadogHook struct {
+	ApiKey          string
+	Tags            *GlobalTags
+	MinLevel        logrus.Level
+	MaxRetry        int
+	datadogEndpoint string
+	entryC          chan logrus.Entry
+	errorC          chan error
+	ticker          *time.Ticker
+	formatter       logrus.Formatter
+	wg              sync.WaitGroup
+}
+
+type Options struct {
+	ApiKey              *string
+	MinimumLoggingLevel *logrus.Level
+}
+
+func getenvOrDefault(key string, defaultString string) string {
+	if val := os.Getenv(key); val != "" {
+		return val
+	}
+
+	return defaultString
+}
+
+func defaultErrorhandler(err error) {
+	logrus.Errorf("The datadog logger hook has encountered an error: %s", err.Error())
```

Comment:
You sometimes use this and sometime use `fmt.Errorf`, I would prefer everything go though an internal error function that uses the `fmt.Errorf`, as if we have encountered an error in the hook we can't really use logrus to log that error.

Reply:
I did think that I was kinda in 2 minds about this, I didn't really know which way to go. I guess I was thinking this hook failing doesn't actually mean logrus has failed and it will still log to any other outputs. E.g. This could still appear in the console, which you can see happening when you run the tests.

I do think it may be better to use `fmt.ErrorF` but that creates an error we don't always have the ability to pass to the calling context.

As it happens the results from the `fmt.ErrorF` calls all end up here anyway.

I guess the question to answer is, should we expose internal errors to the module using this? And, if so, how? I wonder how logrus handles errors, the log functions themselves don't return an error.

Reply:
Looking at other hooks, they usually use `fmt.ErrorF`, so that would be my preference.

Another options could be allowing a configuration option for how internal errors are reported, but I think that we don't really need to do that at the moment.

Reply:
So do you do with the `error` outputted by the `fmt.ErrorF`?

Reply:
Just print it I guess with fmt.Println

Reply:
Updated